### PR TITLE
Use precompiled regex for ID extractor

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
When using regular expressions with the ID extractor, we compile them on every request, following the suggestions from the performance document, this modifies the code in order to compile the expression only once.

To test this, the following manifest can be used:
```json
  "custom_middleware": {
    "auth_check": {
      "name": "MyAuthCheck"
    },
    "id_extractor": {
      "extract_from": "header",
      "extract_with": "regex",
      "extractor_config": {
        "header_name": "Authorization",
        "regex_expression": "prefix(.*)",
        "regex_match_index": 0
      }
    },
    "driver": "grpc"
  }
```

### CODE REVIEW

**Comment by matiasinsaurralde** [comment]
@dencoded good suggestion

**Comment by matiasinsaurralde** [comment]
@dencoded I have refactored it a bit in order to re-use the decoded structure.

File: coprocess_id_extractor.go [hunk 1]
```diff
@@ -164,62 +166,59 @@ type RegexExtractorConfig struct {
...
-	var extractorOutput string
-	var config RegexExtractorConfig
-
-	err := mapstructure.Decode(e.Config.ExtractorConfig, &config)
-	if err != nil {
-		returnOverrides = e.Error(r, err, "Can't decode RegexExtractor configuration")
-		return SessionID, returnOverrides
+	// Parse specific configuration settings:
+	if e.cfg == nil {
+		if err := mapstructure.Decode(e.Config.ExtractorConfig, &e.cfg); err != nil {
```
Comment (dencoded) [inline coprocess_id_extractor.go]: seems like taking address of nil pointer here, we don't need to take address anyways as `e.cfg` is already pointer

Comment (matiasinsaurralde) [reply coprocess_id_extractor.go]: https://github.com/mitchellh/mapstructure/blob/master/mapstructure.go#L194

Comment (dencoded) [reply coprocess_id_extractor.go]: I mean you probably  wanted to do
```
if e.cfg == nil {
  config := &RegexExtractorConfig{}
  if err := mapstructure.Decode(e.Config.ExtractorConfig, config) {
  ...
  }
  e.cfg = config
...
```

Comment (matiasinsaurralde) [reply coprocess_id_extractor.go]: 👍

**Comment by buger** [comment]
@matiasinsaurralde can you put benchcmp results?

**Comment by buger** [comment]
@matiasinsaurralde bump ^

**Comment by matiasinsaurralde** [comment]
Rebased this with #1695 changes, `benchcmp` output looks like this, running with `go test -benchmem -run=^$ . -bench ^BenchmarkRegexExtractor$ -benchtime=10s`:
```
% benchcmp old new
benchmark                                  old ns/op     new ns/op     delta
BenchmarkRegexExtractor/HeaderSource-8     185811        141028        -24.10%
BenchmarkRegexExtractor/BodySource-8       180877        2369          -98.69%
BenchmarkRegexExtractor/FormSource-8       191935        150725        -21.47%

benchmark                                  old allocs     new allocs     delta
BenchmarkRegexExtractor/HeaderSource-8     117            51             -56.41%
BenchmarkRegexExtractor/BodySource-8       124            18             -85.48%
BenchmarkRegexExtractor/FormSource-8       145            79             -45.52%

benchmark                                  old bytes     new bytes     delta
BenchmarkRegexExtractor/HeaderSource-8     48704         5206          -89.31%
BenchmarkRegexExtractor/BodySource-8       52981         1569          -97.04%
BenchmarkRegexExtractor/FormSource-8       55096         11597         -78.95%
```

**Comment by dencoded** [comment]
benchcmp results looks cool!

File: coprocess_id_extractor.go [hunk 2]
```diff
@@ -76,14 +75,8 @@ func (e *BaseExtractor) ExtractForm(r *http.Request, paramName string) (formValu
...
 
-// ExtractBody is used when BodySource is specified.
 func (e *BaseExtractor) ExtractBody(r *http.Request) (bodyValue string, err error) {
-	copiedRequest := copyRequest(r)
-	body, err := ioutil.ReadAll(copiedRequest.Body)
-	if err != nil {
-		return bodyValue, err
-	}
-	return string(body), err
+	return bodyValue, err
```
Comment (dencoded) [inline coprocess_id_extractor.go]: is this change something left from debugging?

**Comment by matiasinsaurralde** [comment]
@dencoded Broke the tests during the rebase, these are the actual results (not bad):
```
% benchcmp old new
benchmark                                  old ns/op     new ns/op     delta
BenchmarkRegexExtractor/HeaderSource-8     177926        144125        -19.00%
BenchmarkRegexExtractor/BodySource-8       178634        143564        -19.63%
BenchmarkRegexExtractor/FormSource-8       187320        150011        -19.92%

benchmark                                  old allocs     new allocs     delta
BenchmarkRegexExtractor/HeaderSource-8     122            57             -53.28%
BenchmarkRegexExtractor/BodySource-8       129            66             -48.84%
BenchmarkRegexExtractor/FormSource-8       150            85             -43.33%

benchmark                                  old bytes     new bytes     delta
BenchmarkRegexExtractor/HeaderSource-8     48942         5446          -88.87%
BenchmarkRegexExtractor/BodySource-8       53216         9786          -81.61%
BenchmarkRegexExtractor/FormSource-8       55331         11836         -78.61%
```

**Comment by dencoded** [comment]
@matiasinsaurralde it still looks cool

**Pull request merged**